src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3888,61 +3905,65 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
             EnterRegion(finallyRegion);
             AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block));
 
-            if (!knownToImplementIDisposable)
-            {
-                Debug.Assert(!isNotNullableValueType(resource.Type));
-                resource = ConvertToIDisposable(resource, iDisposable, isTryCast: true);
-                int captureId = GetNextCaptureId(finallyRegion);
-                AddStatement(new FlowCaptureOperation(captureId, resource.Syntax, resource));
-                resource = GetCaptureReference(captureId, resource);
-            }
-
-            if (!knownToImplementIDisposable || !isNotNullableValueType(resource.Type))
-            {
-                IOperation condition = MakeIsNullOperation(OperationCloner.CloneOperation(resource));
-                ConditionalBranch(condition, jumpIfTrue: true, endOfFinally);
-                _currentBasicBlock = null;
-            }
-
-            if (!iDisposable.Equals(resource.Type))
+            if (disposeMethod is object)
             {
-                resource = ConvertToIDisposable(resource, iDisposable);
-            }
-
-            AddStatement(tryDispose(resource) ??
-                         MakeInvalidOperation(type: null, resource));
+                ITypeSymbol iDisposable = isAsynchronous
+                                          ? _compilation.CommonGetWellKnownType(WellKnownType.System_IAsyncDisposable).GetITypeSymbol()
+                                          : _compilation.GetSpecialType(SpecialType.System_IDisposable);
 
-            AppendNewBlock(endOfFinally);
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.Default);
```

Comment:
>Default [](start = 112, length = 7)

I think it will be more robust long term to use comparer that ignores all possible aspects here.